### PR TITLE
[flang][OpenMP] Concatenate begin and end clauses into single list

### PR DESCRIPTION
**Pull request by kparzysz**
This will remove the distinction between begin clauses and end clauses,
and process all of them together.

### CODE REVIEW

**Review by tblah — APPROVED** [review]
LGTM, thanks!

**Review by skatrak — APPROVED** [review]
Thanks, LGTM!

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 1]
```diff
@@ -1091,16 +1091,12 @@ static void genParallelClauses(
...
-    cp.processAllocate(clauseOps);
-    cp.processSectionsReduction(loc, clauseOps);
-    // TODO Support delayed privatization.
-  } else {
-    cp.processNowait(clauseOps);
-  }
+  cp.processAllocate(clauseOps);
+  cp.processSectionsReduction(loc, clauseOps);
+  // TODO Support delayed privatization.
+  cp.processNowait(clauseOps);
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/OpenMP.cpp]: Done.

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 2]
```diff
@@ -1280,30 +1274,26 @@ static void genWsloopClauses(
...
   // TODO Support delayed privatization.
 
   if (ReductionProcessor::doReductionByRef(clauseOps.reductionVars))
     clauseOps.reductionByRefAttr = firOpBuilder.getUnitAttr();
 
-  if (!endClauses.empty()) {
-    ClauseProcessor ecp(converter, semaCtx, endClauses);
-    ecp.processNowait(clauseOps);
-  }
+  cp.processNowait(clauseOps);
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/OpenMP.cpp]: Done.

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 3]
```diff
@@ -1121,16 +1117,14 @@ static void genSimdClauses(Fortran::lower::AbstractConverter &converter,
...
-  bcp.processAllocate(clauseOps);
+  ClauseProcessor cp(converter, semaCtx, clauses);
+  cp.processAllocate(clauseOps);
   // TODO Support delayed privatization.
 
-  ClauseProcessor ecp(converter, semaCtx, endClauses);
-  ecp.processCopyprivate(loc, clauseOps);
-  ecp.processNowait(clauseOps);
+  cp.processCopyprivate(loc, clauseOps);
+  cp.processNowait(clauseOps);
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1125]: Done.

**Pull request merged**